### PR TITLE
Trace PeerSelectionState

### PR DESCRIPTION
**Pull request by karknu**
# Description

Trace the peerselection state and peermetrics on sig USR1.
Implements #4730 .

# Checklist

- Branch
    - [x] Updated changelog files.
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [ ] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

### CODE REVIEW

File: ouroboros-network/CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 
 ### Breaking changes
 
+Add TraceDebugState message to TracePeerSelection for tracing
+  peer selection upon getting a USR1 sig.
```
Comment (coot) [inline ouroboros-network/CHANGELOG.md:8]:
```suggestion
* Added `TraceDebugState` message to `TracePeerSelection` for tracing
  peer selection upon getting a USR1 sig.
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 2]
```diff
@@ -794,6 +811,8 @@ runM Interfaces
         }
 
       publicStateVar <- newTVarIO emptyPublicPeerSelectionState
+      debugStateVarA <- newTVarIO $ emptyPeerSelectionState fuzzRng []
+      debugStateVarB <- newTVarIO $ emptyPeerSelectionState fuzzRng []
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:815]: I think we can have just one `TVar` we never run both `initiator-only` and `initiator-and-responder` modes at the same time.

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 3]
```diff
@@ -1166,24 +1191,57 @@ run tracers tracersExtra args argsExtra apps appsExtra = do
...
+                                            inProgressBigLedgerPeersReq
+                                        , dpssInProgressPeerShareReqs = inProgressPeerShareReqs
+                                        , dpssInProgressPromoteCold = inProgressPromoteCold
+                                        , dpssInProgressPromoteWarm = inProgressPromoteWarm
+                                        , dpssInProgressDemoteWarm = inProgressDemoteWarm
+                                        , dpssInProgressDemoteHot = inProgressDemoteHot
+                                        , dpssInProgressDemoteToCold = inProgressDemoteToCold
+                                        , dpssUpstreamyness = up
+                                        , dpssFetchynessBlocks = bp
+                                        }
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:1236]: Could you abstract to a pure function:

```
makeDebugPeerSelectionState :: PeerSelectionState -> ... -> DebugPeerSelectionState
...
```

This function should be defined next to `DebugPeerSelectionState` type.

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 4]
```diff
@@ -1166,24 +1191,57 @@ run tracers tracersExtra args argsExtra apps appsExtra = do
...
 #ifdef POSIX
-                 diInstallSigUSR1Handler = \connectionManager -> do
+                 diInstallSigUSR1Handler = \connectionManager dbgStateVar metrics -> do
                    _ <- Signals.installHandler
                      Signals.sigUSR1
                      (Signals.Catch
                        (do state <- atomically $ readState connectionManager
                            traceWith (dtConnectionManagerTracer tracersExtra)
                                      (TrState state)
+                           PeerSelectionState{..} <- atomically $ readTVar dbgStateVar
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:1208]:
```suggestion
                           PeerSelectionState{..} <- readTVarIO dbgStateVar
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs [hunk 5]
```diff
@@ -1166,24 +1191,57 @@ run tracers tracersExtra args argsExtra apps appsExtra = do
...
+                           now <- getMonotonicTime
+                           up <- atomically $ upstreamyness metrics
+                           bp <- atomically $ fetchynessBlocks metrics
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:1211]:
```suggestion
                           (up, bp) <- atomically $ (,) <*> upstreamyness metrics
                                                        <$> fetchynessBlocks metrics
```

**Review by coot — CHANGES_REQUESTED** [review]
Nice, just a few minor comments.

File: ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs (hunk 2 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs:815]: They are two different variables because they are different types.

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Diffusion/P2P.hs]: I've moved the creation of the variables so it is only one variable is created depending on mode.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs [hunk 6]
```diff
@@ -437,6 +439,64 @@ data PeerSelectionState peeraddr peerconn = PeerSelectionState {
...
+       dpssInProgressBigLedgerPeersReq :: !Bool,
+       dpssInProgressPeerShareReqs     :: !Int,
+       dpssInProgressPromoteCold       :: !(Set peeraddr),
+       dpssInProgressPromoteWarm       :: !(Set peeraddr),
+       dpssInProgressDemoteWarm        :: !(Set peeraddr),
+       dpssInProgressDemoteHot         :: !(Set peeraddr),
+       dpssInProgressDemoteToCold      :: !(Set peeraddr),
+       dpssUpstreamyness               :: !(Map peeraddr Int),
+       dpssFetchynessBlocks            :: !(Map peeraddr Int)
+} deriving Show
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/Governor/Types.hs:469]: Tracing types have be lazy (that's a requirement of the new tracing system). Most of the computations done in `makeDebugPeerSelectionState` are trivial so it doesn't matter much for performance.

**Pull request merged**